src/xwayland/xserver.rs

```diff
@@ -65,167 +66,225 @@ use wayland_server::{Client, Display, Filter};
-pub trait XWindowManager {
+/// Your WM code must be able handle the XWayland server connecting then
+/// disconnecting several time in a row, but only a single connection will
+/// be active at any given time.
+pub enum XWaylandEvent {
     /// The XWayland server is ready
-    ///
-    /// Your privileged connection to it is this `UnixStream`
-    fn xwayland_ready(&mut self, connection: UnixStream, client: Client);
-    /// The XWayland server has exited
-    fn xwayland_exited(&mut self);
+    Ready {
+        /// Privileged X11 connection to XWayland
+        connection: UnixStream,
+        /// Wayland client representing XWayland
+        client: Client,
+    },
+    /// The XWayland server exited
+    Exited,
 }
 
-impl<WM: XWindowManager + 'static> XWayland<WM> {
-    /// Start the XWayland server
-    pub fn init<L, T: Any, Data: 'static>(
-        wm: WM,
+impl<Data: Any + 'static> XWayland<Data> {
+    /// Create a new XWayland manager
+    pub fn new<L>(
         handle: LoopHandle<'static, Data>,
         display: Rc<RefCell<Display>>,
-        data: &mut T,
         logger: L,
-    ) -> Result<XWayland<WM>, ()>
+    ) -> (XWayland<Data>, XWaylandSource)
     where
         L: Into<Option<::slog::Logger>>,
     {
         let log = crate::slog_or_fallback(logger);
+        // We don't expect to ever have more than 2 messages in flight, if XWayland got ready and then died right away
+        let (sender, channel) = sync_channel(2);
         let inner = Rc::new(RefCell::new(Inner {
-            wm,
-            kill_source: {
-                let handle = handle.clone();
-                Box::new(move |source| handle.kill(source))
-            },
-            source_maker: Box::new(move |inner, fd| {
-                handle
-                    .insert_source(
-                        Generic::new(Fd(fd), Interest::READ, Mode::Level),
-                        move |evt, _, _| {
-                            debug_assert!(evt.readable);
-                            xwayland_ready(&inner);
-                            Ok(())
-                        },
-                    )
-                    .map_err(|_| ())
-            }),
+            handle,
             wayland_display: display,
             instance: None,
+            sender,
             log: log.new(o!("smithay_module" => "XWayland")),
         }));
-        launch(&inner, data)?;
-        Ok(XWayland { inner })
+        (XWayland { inner }, XWaylandSource { channel })
+    }
+
+    /// Attempt to start the XWayland instance
+    ///
+    /// If it succeeds, you'll eventually receive an `XWaylandEvent::Ready`
+    /// through the source provided by `XWayland::new()` containing an
+    /// `UnixStream` representing your WM connection to XWayland, and the
+    /// wayland `Client` for XWayland.
+    ///
+    /// Does nothing if XWayland is already started or starting.
+    ///
+    /// You need to provide this function the same `DispatchData` you pass to the main
```

Comment:
Where do I provide this function the `DispatchData`?

Reply:
Ah indeed, thanks, this is a leftover from an intermediate state of the refactor. :sweat_smile: